### ISSUE TITLE
Different Prediction Probabilities from Extracted Workflow Fit vs. Manually Created Workflow Fit vs. `last_fit`

### ISSUE THREAD

**Issue by AndrewKostandy** [tidymodels/tune#443]
## The problem

Thank you for all your great work on Tidymodels.

I noticed that using `fit` then `predict` on an extracted workflow from a workflow set generates different probability predictions on the test set than when using `fit` then `predict` on a manually created identical workflow (using the same hyperparameters) and both of those predictions are different than the predictions generated from `last_fit` on both workflows resulting in 4 different sets of predictions. Is there a way to ensure identical predictions for all by setting seeds somewhere?

Specifically: 

Prediction Method 1) Extract a workflow from a workflow set then finalize the workflow with the best hyperparameters, fit on training data, then predict on test data

Prediction Method 2) Build an identical workflow from scratch using the best hyperparameters identified earlier, fit  on training data, then predict on test data

Prediction Method 3A) Use `last_fit` on the manually built workflow from prediction method 2

Prediction Method 3B)  Use `last_fit` on the extracted & finalized workflow from prediction method 1.

These 4 methods result in 4 different prediction probability values which are close but not identical.

This issue is similar to a previous issue I had opened some time back [here](https://github.com/tidymodels/tune/issues/300) which was fixed.

## Reproducible example

``` r
library(tidyverse)
suppressPackageStartupMessages(library(tidymodels))
#> Warning: package 'broom' was built under R version 4.1.2
library(finetune)
library(modeldata)
suppressPackageStartupMessages(library(janitor))
suppressPackageStartupMessages(library(GGally))
suppressPackageStartupMessages(library(doMC))

data(attrition)

attrition <- clean_names(attrition)
attrition <- mutate(attrition, attrition = fct_rev(attrition)) %>%
  rename(attrition_response = attrition)

registerDoMC(cores = detectCores()-2)

set.seed(1954)
attrition_split <- initial_split(attrition, strata = "attrition_response")
attrition_train <- training(attrition_split)
attrition_test <- testing(attrition_split)

set.seed(2847)
folds <- vfold_cv(attrition_train, v = 10, strata = "attrition_response")

my_metrics <- metric_set(roc_auc)

rf_recipe <- 
  recipe(attrition_response ~ ., data = attrition_train) %>% 
  step_novel(all_nominal_predictors()) %>%
  step_other(all_nominal_predictors(), threshold = tune()) %>% 
  step_zv(all_predictors())

rf_spec <- rand_forest(mtry = tune(), trees = tune(), min_n = tune()) %>% 
  set_mode("classification") %>% 
  set_engine("ranger")

set.seed(85)
rf_grid <- grid_latin_hypercube(
  mtry() %>% range_set(c(1, 30)),
  min_n(),
  trees(),
  threshold(),
  size = 60
)

# To shorten tuning time
rf_grid <- slice(rf_grid, 33:42)

rec_list <- list(rf_recipe = rf_recipe)
mod_list <- list(rf_spec = rf_spec)

wf_set <- workflow_set(rec_list, mod_list)

wf_set <- wf_set %>%
  option_add(grid = rf_grid, id = "rf_recipe_rf_spec")

results_wflow_set <- workflow_map(
  wf_set,
  fn = "tune_race_anova",
  resamples = folds,
  metrics = my_metrics,
  seed = 85
)

best_hyperparameters <- results_wflow_set %>% 
  extract_workflow_set_result("rf_recipe_rf_spec") %>% 
  select_best("roc_auc")

rf_extracted_workflow <-
  results_wflow_set %>% 
  extract_workflow("rf_recipe_rf_spec") %>% 
  finalize_workflow(best_hyperparameters)

# Predition Method 1 (from finalized workflow)
rf_extracted_wf_predictions <- rf_extracted_workflow %>% 
  fit(attrition_train) %>% 
  predict(attrition_test, type = "prob")


# Next, let's fit the model ourselves using the best hyperparameters
rf_recipe_manual <- 
  recipe(attrition_response ~ ., data = attrition_train) %>% 
  step_novel(all_nominal_predictors()) %>%
  step_other(all_nominal_predictors(), threshold = best_hyperparameters$threshold) %>% 
  step_zv(all_predictors())

rf_spec_manual <- rand_forest(mtry = best_hyperparameters$mtry,
                       trees = best_hyperparameters$trees,
                       min_n = best_hyperparameters$min_n) %>% 
  set_mode("classification") %>% 
  set_engine("ranger")

rf_manual_workflow <-
  workflow() %>% 
  add_recipe(rf_recipe_manual) %>% 
  add_model(rf_spec_manual)


rf_manual_fit_on_train <- fit(rf_manual_workflow, attrition_train)

# Predition Method 2 (from fit of 'rf_workflow')
rf_manual_wf_predictions <- predict(rf_manual_fit_on_train, attrition_test, type = "prob")

# Next, let's use last_fit on 'rf_workflow'
rf_last_fit_manual_wf <- last_fit(rf_manual_workflow, attrition_split)

# Predition Method 3A (from last_fit on 'rf_workflow')
last_fit_manual_wf_predictions <- rf_last_fit_manual_wf %>% collect_predictions()

# Next, let's use last_fit on the extracted and finalized workflow
rf_last_fit_extracted_wf <- last_fit(rf_extracted_workflow, attrition_split)

# Prediction Method 3B (from last_fit on `rf_extracted_workflow`)
last_fit_extracted_wf_predictions <- rf_last_fit_extracted_wf %>% collect_predictions()


various_preds <- tibble(
  extracted_wf = rf_extracted_wf_predictions$.pred_Yes, # Prediction Method 1
  manual_wf = rf_manual_wf_predictions$.pred_Yes,       # Prediction Method 2
  last_fit_manual_wf = last_fit_manual_wf_predictions$.pred_Yes,   # Prediction Method 3A
  last_fit_extracted_wf = last_fit_extracted_wf_predictions$.pred_Yes # Prediction Method 3B
)

various_preds
#> # A tibble: 369 × 4
#>    extracted_wf manual_wf last_fit_manual_wf last_fit_extracted_wf
#>           <dbl>     <dbl>              <dbl>                 <dbl>
#>  1       0.232     0.227              0.229                 0.219 
#>  2       0.359     0.352              0.354                 0.355 
#>  3       0.230     0.213              0.229                 0.227 
#>  4       0.387     0.392              0.384                 0.381 
#>  5       0.0863    0.0865             0.0908                0.0835
#>  6       0.208     0.201              0.207                 0.197 
#>  7       0.184     0.191              0.193                 0.206 
#>  8       0.487     0.501              0.488                 0.506 
#>  9       0.0905    0.0813             0.0810                0.0864
#> 10       0.105     0.118              0.105                 0.112 
#> # … with 359 more rows

ggpairs(various_preds)
```

![](https://i.imgur.com/zhDwIKV.png)

<sup>Created on 2022-01-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.1.1 (2021-08-10)
#>  os       macOS Big Sur 10.16
#>  system   x86_64, darwin17.0
#>  ui       X11
#>  language (EN)
#>  collate  en_CA.UTF-8
#>  ctype    en_CA.UTF-8
#>  tz       America/Toronto
#>  date     2022-01-08
#>  pandoc   2.14.0.3 @ /Applications/RStudio.app/Contents/MacOS/pandoc/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package      * version    date (UTC) lib source
#>  assertthat     0.2.1      2019-03-21 [1] CRAN (R 4.1.0)
#>  backports      1.4.1      2021-12-13 [1] CRAN (R 4.1.0)
#>  boot           1.3-28     2021-05-03 [1] CRAN (R 4.1.1)
#>  broom        * 0.7.11     2022-01-03 [1] CRAN (R 4.1.2)
#>  cellranger     1.1.0      2016-07-27 [1] CRAN (R 4.1.0)
#>  class          7.3-19     2021-05-03 [1] CRAN (R 4.1.1)
#>  cli            3.1.0      2021-10-27 [1] CRAN (R 4.1.0)
#>  codetools      0.2-18     2020-11-04 [1] CRAN (R 4.1.1)
#>  colorspace     2.0-2      2021-06-24 [1] CRAN (R 4.1.0)
#>  crayon         1.4.2      2021-10-29 [1] CRAN (R 4.1.0)
#>  curl           4.3.2      2021-06-23 [1] CRAN (R 4.1.0)
#>  DBI            1.1.2      2021-12-20 [1] CRAN (R 4.1.0)
#>  dbplyr         2.1.1      2021-04-06 [1] CRAN (R 4.1.0)
#>  dials        * 0.0.10     2021-09-10 [1] CRAN (R 4.1.1)
#>  DiceDesign     1.9        2021-02-13 [1] CRAN (R 4.1.0)
#>  digest         0.6.29     2021-12-01 [1] CRAN (R 4.1.1)
#>  doMC         * 1.3.7      2020-10-14 [1] CRAN (R 4.1.0)
#>  dplyr        * 1.0.7      2021-06-18 [1] CRAN (R 4.1.0)
#>  ellipsis       0.3.2      2021-04-29 [1] CRAN (R 4.1.0)
#>  evaluate       0.14       2019-05-28 [1] CRAN (R 4.1.0)
#>  fansi          0.5.0      2021-05-25 [1] CRAN (R 4.1.0)
#>  farver         2.1.0      2021-02-28 [1] CRAN (R 4.1.0)
#>  fastmap        1.1.0      2021-01-25 [1] CRAN (R 4.1.0)
#>  finetune     * 0.1.0      2021-07-21 [1] CRAN (R 4.1.0)
#>  forcats      * 0.5.1      2021-01-27 [1] CRAN (R 4.1.0)
#>  foreach      * 1.5.1      2020-10-15 [1] CRAN (R 4.1.0)
#>  fs             1.5.2      2021-12-08 [1] CRAN (R 4.1.0)
#>  furrr          0.2.3      2021-06-25 [1] CRAN (R 4.1.0)
#>  future         1.23.0     2021-10-31 [1] CRAN (R 4.1.0)
#>  future.apply   1.8.1      2021-08-10 [1] CRAN (R 4.1.0)
#>  generics       0.1.1      2021-10-25 [1] CRAN (R 4.1.0)
#>  GGally       * 2.1.2      2021-06-21 [1] CRAN (R 4.1.0)
#>  ggplot2      * 3.3.5      2021-06-25 [1] CRAN (R 4.1.0)
#>  globals        0.14.0     2020-11-22 [1] CRAN (R 4.1.0)
#>  glue           1.6.0      2021-12-17 [1] CRAN (R 4.1.1)
#>  gower          0.2.2      2020-06-23 [1] CRAN (R 4.1.0)
#>  GPfit          1.0-8      2019-02-08 [1] CRAN (R 4.1.0)
#>  gtable         0.3.0      2019-03-25 [1] CRAN (R 4.1.0)
#>  hardhat        0.1.6      2021-07-14 [1] CRAN (R 4.1.0)
#>  haven          2.4.3      2021-08-04 [1] CRAN (R 4.1.0)
#>  highr          0.9        2021-04-16 [1] CRAN (R 4.1.0)
#>  hms            1.1.1      2021-09-26 [1] CRAN (R 4.1.0)
#>  htmltools      0.5.2      2021-08-25 [1] CRAN (R 4.1.0)
#>  httr           1.4.2      2020-07-20 [1] CRAN (R 4.1.0)
#>  infer        * 1.0.0      2021-08-13 [1] CRAN (R 4.1.0)
#>  ipred          0.9-12     2021-09-15 [1] CRAN (R 4.1.0)
#>  iterators    * 1.0.13     2020-10-15 [1] CRAN (R 4.1.0)
#>  janitor      * 2.1.0      2021-01-05 [1] CRAN (R 4.1.0)
#>  jsonlite       1.7.2      2020-12-09 [1] CRAN (R 4.1.0)
#>  knitr          1.37       2021-12-16 [1] CRAN (R 4.1.1)
#>  labeling       0.4.2      2020-10-20 [1] CRAN (R 4.1.0)
#>  lattice        0.20-45    2021-09-22 [1] CRAN (R 4.1.0)
#>  lava           1.6.10     2021-09-02 [1] CRAN (R 4.1.0)
#>  lhs            1.1.3      2021-09-08 [1] CRAN (R 4.1.0)
#>  lifecycle      1.0.1      2021-09-24 [1] CRAN (R 4.1.0)
#>  listenv        0.8.0      2019-12-05 [1] CRAN (R 4.1.0)
#>  lme4           1.1-27.1   2021-06-22 [1] CRAN (R 4.1.0)
#>  lubridate      1.8.0      2021-10-07 [1] CRAN (R 4.1.0)
#>  magrittr       2.0.1      2020-11-17 [1] CRAN (R 4.1.0)
#>  MASS           7.3-54     2021-05-03 [1] CRAN (R 4.1.1)
#>  Matrix         1.4-0      2021-12-08 [1] CRAN (R 4.1.0)
#>  mime           0.12       2021-09-28 [1] CRAN (R 4.1.0)
#>  minqa          1.2.4      2014-10-09 [1] CRAN (R 4.1.0)
#>  modeldata    * 0.1.1      2021-07-14 [1] CRAN (R 4.1.0)
#>  modelr         0.1.8      2020-05-19 [1] CRAN (R 4.1.0)
#>  munsell        0.5.0      2018-06-12 [1] CRAN (R 4.1.0)
#>  nlme           3.1-153    2021-09-07 [1] CRAN (R 4.1.0)
#>  nloptr         1.2.2.3    2021-11-02 [1] CRAN (R 4.1.0)
#>  nnet           7.3-16     2021-05-03 [1] CRAN (R 4.1.1)
#>  parallelly     1.30.0     2021-12-17 [1] CRAN (R 4.1.0)
#>  parsnip      * 0.1.7      2021-07-21 [1] CRAN (R 4.1.0)
#>  pillar         1.6.4      2021-10-18 [1] CRAN (R 4.1.0)
#>  pkgconfig      2.0.3      2019-09-22 [1] CRAN (R 4.1.0)
#>  plyr           1.8.6      2020-03-03 [1] CRAN (R 4.1.0)
#>  pROC           1.18.0     2021-09-03 [1] CRAN (R 4.1.0)
#>  prodlim        2019.11.13 2019-11-17 [1] CRAN (R 4.1.0)
#>  purrr        * 0.3.4      2020-04-17 [1] CRAN (R 4.1.0)
#>  R.cache        0.15.0     2021-04-30 [1] CRAN (R 4.1.0)
#>  R.methodsS3    1.8.1      2020-08-26 [1] CRAN (R 4.1.0)
#>  R.oo           1.24.0     2020-08-26 [1] CRAN (R 4.1.0)
#>  R.utils        2.11.0     2021-09-26 [1] CRAN (R 4.1.0)
#>  R6             2.5.1      2021-08-19 [1] CRAN (R 4.1.0)
#>  ranger       * 0.13.1     2021-07-14 [1] CRAN (R 4.1.0)
#>  RColorBrewer   1.1-2      2014-12-07 [1] CRAN (R 4.1.0)
#>  Rcpp           1.0.7      2021-07-07 [1] CRAN (R 4.1.0)
#>  readr        * 2.1.1      2021-11-30 [1] CRAN (R 4.1.0)
#>  readxl         1.3.1      2019-03-13 [1] CRAN (R 4.1.0)
#>  recipes      * 0.1.17     2021-09-27 [1] CRAN (R 4.1.0)
#>  reprex         2.0.1      2021-08-05 [1] CRAN (R 4.1.0)
#>  reshape        0.8.8      2018-10-23 [1] CRAN (R 4.1.0)
#>  rlang        * 0.4.12     2021-10-18 [1] CRAN (R 4.1.0)
#>  rmarkdown      2.11       2021-09-14 [1] CRAN (R 4.1.0)
#>  rpart          4.1-15     2019-04-12 [1] CRAN (R 4.1.1)
#>  rsample      * 0.1.1      2021-11-08 [1] CRAN (R 4.1.0)
#>  rstudioapi     0.13       2020-11-12 [1] CRAN (R 4.1.0)
#>  rvest          1.0.2      2021-10-16 [1] CRAN (R 4.1.0)
#>  scales       * 1.1.1      2020-05-11 [1] CRAN (R 4.1.0)
#>  sessioninfo    1.2.2      2021-12-06 [1] CRAN (R 4.1.0)
#>  snakecase      0.11.0     2019-05-25 [1] CRAN (R 4.1.0)
#>  stringi        1.7.6      2021-11-29 [1] CRAN (R 4.1.0)
#>  stringr      * 1.4.0      2019-02-10 [1] CRAN (R 4.1.0)
#>  styler         1.6.2      2021-09-23 [1] CRAN (R 4.1.0)
#>  survival       3.2-13     2021-08-24 [1] CRAN (R 4.1.0)
#>  tibble       * 3.1.6      2021-11-07 [1] CRAN (R 4.1.0)
#>  tidymodels   * 0.1.4      2021-10-01 [1] CRAN (R 4.1.0)
#>  tidyr        * 1.1.4      2021-09-27 [1] CRAN (R 4.1.0)
#>  tidyselect     1.1.1      2021-04-30 [1] CRAN (R 4.1.0)
#>  tidyverse    * 1.3.1      2021-04-15 [1] CRAN (R 4.1.0)
#>  timeDate       3043.102   2018-02-21 [1] CRAN (R 4.1.0)
#>  tune         * 0.1.6      2021-07-21 [1] CRAN (R 4.1.0)
#>  tzdb           0.2.0      2021-10-27 [1] CRAN (R 4.1.0)
#>  utf8           1.2.2      2021-07-24 [1] CRAN (R 4.1.0)
#>  vctrs        * 0.3.8      2021-04-29 [1] CRAN (R 4.1.0)
#>  withr          2.4.3      2021-11-30 [1] CRAN (R 4.1.0)
#>  workflows    * 0.2.4      2021-10-12 [1] CRAN (R 4.1.0)
#>  workflowsets * 0.1.0      2021-07-22 [1] CRAN (R 4.1.0)
#>  xfun           0.29       2021-12-14 [1] CRAN (R 4.1.0)
#>  xml2           1.3.3      2021-11-30 [1] CRAN (R 4.1.0)
#>  yaml           2.2.1      2020-02-01 [1] CRAN (R 4.1.0)
#>  yardstick    * 0.0.9      2021-11-22 [1] CRAN (R 4.1.0)
#> 
#>  [1] /Library/Frameworks/R.framework/Versions/4.1/Resources/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>
